### ISSUE TITLE
BUG: bar deals with bytes and string x data in different manners, both that are unexpected

### ISSUE THREAD

**Issue by NelleV** [matplotlib/matplotlib#7215]
And messes quite badly…

```python
import matplotlib.pyplot as plt
import numpy as np

types = np.array([b"1.0", b"11.0", b"6.0"])
counts = np.array([544, 631,  55])
fig, axes = plt.subplots(ncols=2)
axes[0].bar(types, counts)

types = np.array(["1.0", "11.0", "6.0"])
counts = np.array([544, 631,  55])
axes[1].bar(types, counts)
```

yields (on matplotlib

![mpl_bug_bar_1](https://cloud.githubusercontent.com/assets/184798/19056850/8632ff96-8980-11e6-9e2a-cc8523989416.png)

on matplotlib version "2.0.0b4+2254.gf544a83"

Neither of those plots are correct.

**Comment by tacaswell** [comment]
This come up as part of the categorical work we decided to silently kill some of this 'feature', on master you get:

![so](https://cloud.githubusercontent.com/assets/199813/19058821/899db3d8-89a8-11e6-830c-43a5032aeb57.png)

but 

```python
In [8]: import matplotlib.pyplot as plt
   ...: import numpy as np
   ...: 
   ...: types = [b"1.0", b"11.0", b"6.0"]
   ...: counts = np.array([544, 631,  55])
   ...: fig, axes = plt.subplots(ncols=2)
   ...: axes[0].bar(types, counts)
   ...: 
   ...: types = np.array(["1.0", "11.0", "6.0"])
   ...: counts = np.array([544, 631,  55])
   ...: axes[1].bar(types, counts)
   ...: 
   ...: 

```

Not clear what is going wrong with the array of bytes.

**Comment by tacaswell** [comment]
Sorry, I miss-read, thought this was on 2.x

My image is from '2.0.0b4.post2369+gb1bab75'

**Comment by tacaswell** [comment]
attn @story645

**Comment by NelleV** [comment]
@tacaswell so you are getting what I expect to get. I may have not updated my code recently enough to have the last updates. I am also wondering whether I uncovered one or two bugs. But maybe @story645 can shed some light on this.

**Comment by tacaswell** [comment]
And to be clear, what I mean by 'going wrong with bytes', I mean it raises

```pytb
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-3-f8a9e716484b> in <module>()
      5 counts = np.array([544, 631,  55])
      6 fig, axes = plt.subplots(ncols=2)
----> 7 axes[0].bar(types, counts)
      8 
      9 types = np.array(["1.0", "11.0", "6.0"])

/home/tcaswell/source/p/matplotlib/lib/matplotlib/__init__.py in inner(ax, *args, **kwargs)
   1859                     warnings.warn(msg % (label_namer, func.__name__),
   1860                                   RuntimeWarning, stacklevel=2)
-> 1861             return func(ax, *args, **kwargs)
   1862         pre_doc = inner.__doc__
   1863         if pre_doc is None:

/home/tcaswell/source/p/matplotlib/lib/matplotlib/axes/_axes.py in bar(self, left, height, width, bottom, **kwargs)
   2111         if align == 'center':
   2112             if orientation == 'vertical':
-> 2113                 left = [left[i] - width[i] / 2. for i in xrange(len(left))]
   2114             elif orientation == 'horizontal':
   2115                 bottom = [bottom[i] - height[i] / 2.

/home/tcaswell/source/p/matplotlib/lib/matplotlib/axes/_axes.py in <listcomp>(.0)
   2111         if align == 'center':
   2112             if orientation == 'vertical':
-> 2113                 left = [left[i] - width[i] / 2. for i in xrange(len(left))]
   2114             elif orientation == 'horizontal':
   2115                 bottom = [bottom[i] - height[i] / 2.

TypeError: unsupported operand type(s) for -: 'numpy.bytes_' and 'float'


```

**Comment by NelleV** [comment]
Hum… Interesting: that's approximatively what I get on 1.5.
I'd actually be happy with that (raising an error, however obscure the error might be, is better than unexpected).

**Comment by NelleV** [comment]
Confirming that this is what I get with master.
Sorry about the confusion: I should have checked I had the latest version before opening the ticket.

**Comment by story645** [comment]
I'm confused, 'cause I'm synced to master ('2.0.0b3.post2557+gb1bab75') and using this code:
```
import matplotlib.pyplot as plt
import numpy as np
types = [b"1.0", b"11.0", b"6.0"]
counts = np.array([544, 631,  55])
fig, axes = plt.subplots(ncols=2)
axes[0].bar(types, counts)

types = np.array(["1.0", "11.0", "6.0"])
counts = np.array([544, 631,  55])
axes[1].bar(types, counts)
```
get this:
![index](https://cloud.githubusercontent.com/assets/1300499/19097732/fa9eb11e-8a74-11e6-88da-433fec42aa8e.png)

@NelleV what version of numpy do you have?

**Comment by story645** [comment]
@tacaswell and @QuLogic, that error message indicates that the conversion isn't getting registered because left [should have been converted to an integer](https://github.com/matplotlib/matplotlib/blob/master/lib/matplotlib/axes/_axes.py#L2093)
I can't replicate this error, but my guess is that adding some types to [category](https://github.com/matplotlib/matplotlib/blob/master/lib/matplotlib/category.py#L134) might fix it.

**Comment by story645** [comment]
Further discussion with @QuLogic indicates that the fix is probably to add the following line to the bottom of [category.py](https://github.com/matplotlib/matplotlib/blob/master/lib/matplotlib/category.py#L134) 
```python
units.registry[np.bytes_] = StrCategoryConverter()
```
Any chance @NelleV or @QuLogic can test it since I can't replicate the error?

**Comment by QuLogic** [comment]
This issue kind of branched topics a bit, but the crashing should be fixed by #7241. I'm not sure if that's enough to close out this bug too.
